### PR TITLE
chore: vitest + react testing library (part 2)

### PR DESCRIPTION
**Pull request by IrakliJani**
Closes: https://github.com/lightdash/lightdash/issues/8739

### Description:

- uses `isomorphic-fetch` to make `lightdashApi` work both in tests and frontend.
- replaces `vitest-fetch-mock` with `nock` and rewrites `api.test.ts` to use `nock`.
- adds provider mocks.
- adds app provider mock and an ability to override health | user properties.
- adds `renderWithProviders` and `renderHookWithProviders`
- adds tests for `useApp` hook as a sample
- adds tests for `UserCompletionModal.test.tsx` as a sample

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/frontend/src/api.ts [hunk 1]
```diff
@@ -4,8 +4,14 @@ import {
     LightdashRequestMethodHeader,
     RequestMethod,
 } from '@lightdash/common';
+import fetch from 'isomorphic-fetch';
 
-const apiPrefix = '/api/v1';
+export const BASE_API_URL =
+    import.meta.env.VITEST === 'true'
+        ? `http://test.lightdash/`
```
Comment (IrakliJani) [inline packages/frontend/src/api.ts:11]: `nock` and `fetch` (node's fetch) fail if url starts without protocol + domain

File: packages/frontend/src/components/UserCompletionModal/UserCompletionModal.test.tsx [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import { describe, expect, it } from 'vitest';
+import { renderWithProviders } from '../../testing/testUtils';
+import UserCompletionModal from './UserCompletionModal';
+
+describe('UserCompletionModal', () => {
+    it("should not render anything if user's setup is complete", async () => {
+        const { container } = renderWithProviders(<UserCompletionModal />);
+
+        await waitFor(() => expect(container).toBeEmptyDOMElement());
+    });
```
Comment (IrakliJani) [inline packages/frontend/src/components/UserCompletionModal/UserCompletionModal.test.tsx:11]: by default `renderWithProviders` mocks user's data as a regular logged in user with `isSetupComplete = true` which makes `<UserCompletionModal />` not to render

File: packages/frontend/src/components/UserCompletionModal/UserCompletionModal.test.tsx [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+    it("should render user completion modal if user's setup is not complete", async () => {
+        const { getByText } = renderWithProviders(<UserCompletionModal />, {
+            user: {
+                isSetupComplete: false,
+            },
+        });
```
Comment (IrakliJani) [inline packages/frontend/src/components/UserCompletionModal/UserCompletionModal.test.tsx:18]: `renderWithProviders` accepts second argument of mocks that can easily override user and health data

File: packages/frontend/src/components/UserCompletionModal/UserCompletionModal.test.tsx [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+        await waitFor(() =>
+            expect(getByText('Nearly there...')).toBeInTheDocument(),
+        );
```
Comment (IrakliJani) [inline packages/frontend/src/components/UserCompletionModal/UserCompletionModal.test.tsx:22]: now that `isSetupComplete` is set to false - modal renders

File: packages/frontend/src/hooks/cartesianChartConfig/useCartesianChartConfig.test.ts [hunk 5]
```diff
@@ -1,5 +1,5 @@
 import { CartesianSeriesType, getItemMap } from '@lightdash/common';
-import { renderHook } from '@testing-library/react-hooks';
```
Comment (IrakliJani) [inline packages/frontend/src/hooks/cartesianChartConfig/useCartesianChartConfig.test.ts:2]: this library is not necessary anymore

File: packages/frontend/src/hooks/toaster/useToaster.tsx [hunk 6]
```diff
@@ -58,7 +58,7 @@ const useToaster = () => {
                     subtitle || action ? (
                         <Stack spacing="xs" align="flex-start">
                             {typeof subtitle == 'string' ? (
-                                <MDEditor.Markdown
+                                <MarkdownPreview
```
Comment (IrakliJani) [inline packages/frontend/src/hooks/toaster/useToaster.tsx:61]: `MarkdownEditor` uses `MarkdownPreview` to render markdown.
lot of react-query hooks use toasters so `MarkdownPreview` is mocked in tests by default

File: packages/frontend/src/components/UserCompletionModal/UserCompletionModal.test.tsx [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import { waitFor } from '@testing-library/react';
+import { describe, expect, it } from 'vitest';
+import { renderWithProviders } from '../../testing/testUtils';
+import UserCompletionModal from './UserCompletionModal';
+
+describe('UserCompletionModal', () => {
+    it("should not render anything if user's setup is complete", async () => {
+        const { container } = renderWithProviders(<UserCompletionModal />);
```
Comment (IrakliJani) [inline packages/frontend/src/components/UserCompletionModal/UserCompletionModal.test.tsx:8]: `renderWithProviders` is a wrapper around testing-library's `render` which renders components inside providers

File: packages/frontend/src/providers/AppProvider.test.tsx [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import { renderHook, waitFor } from '@testing-library/react';
+import { describe, expect, it } from 'vitest';
+import { renderHookWithProviders } from '../testing/testUtils';
+import { useApp } from './AppProvider';
+
+describe('AppProvider', () => {
+    it('should throw if used outside of AppProvider', () => {
+        expect(() => {
+            renderHook(() => useApp());
+        }).toThrow('useApp must be used within a AppProvider');
```
Comment (IrakliJani) [inline packages/frontend/src/providers/AppProvider.test.tsx:10]: `renderHook` does not render hook inside providers. that's why it fails

File: packages/frontend/src/providers/AppProvider.test.tsx [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    });
+
+    it('should return values from AppProvider', () => {
+        expect(async () => {
+            const { result } = renderHookWithProviders(() => useApp());
```
Comment (IrakliJani) [inline packages/frontend/src/providers/AppProvider.test.tsx:15]: `renderHookWithProviders` is very similar to `renderWithProviders`. it also accepts overrides for mocks (user, health) as the second argument!

File: packages/frontend/src/providers/ReactQueryProvider.tsx [hunk 10]
```diff
@@ -0,0 +1,45 @@
+import {
+    DefaultOptions,
+    QueryClient,
+    QueryClientProvider,
+} from '@tanstack/react-query';
+import { FC, PropsWithChildren } from 'react';
+
+// used in test mocks
+// ts-unused-exports:disable-next-line
+export const createQueryClient = (options?: DefaultOptions) => {
```
Comment (IrakliJani) [inline packages/frontend/src/providers/ReactQueryProvider.tsx:10]: this config is our default configuration. I use this + overrides for the mocked `ReactQueryProvider`

File: packages/frontend/src/providers/__mocks__/ReactQueryProvider.tsx [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+import { createQueryClient } from '../ReactQueryProvider';
+
+const ReactQueryProvider: FC<PropsWithChildren> = ({ children }) => {
+    const queryClient = createQueryClient({
+        queries: {
+            retry: false,
+        },
+        mutations: {
+            retry: false,
+        },
```
Comment (IrakliJani) [inline packages/frontend/src/providers/__mocks__/ReactQueryProvider.tsx:12]: overrides necessary for testing https://tanstack.com/query/v4/docs/framework/react/guides/testing#turn-off-retries

File: packages/frontend/src/testing/__mocks__/api/healthResponse.mock.ts [hunk 12]
```diff
@@ -0,0 +1,71 @@
+import { HealthState, LightdashMode } from '@lightdash/common';
+
+export default function mockHealthResponse(
```
Comment (IrakliJani) [inline packages/frontend/src/testing/__mocks__/api/healthResponse.mock.ts:3]: mocked health response

File: packages/frontend/src/testing/__mocks__/api/userResponse.mock.ts [hunk 13]
```diff
@@ -0,0 +1,230 @@
+import {
+    LightdashUserWithAbilityRules,
+    OrganizationMemberRole,
+} from '@lightdash/common';
+
+export function mockUserResponse(
```
Comment (IrakliJani) [inline packages/frontend/src/testing/__mocks__/api/userResponse.mock.ts:6]: mocked user response

File: packages/frontend/src/testing/__mocks__/implementations/matchMedia.mock.ts [hunk 14]
```diff
@@ -0,0 +1,18 @@
+import { vi } from 'vitest';
+
+function mockMatchMedia() {
+    window.matchMedia = vi.fn().mockImplementation((query) => {
```
Comment (IrakliJani) [inline packages/frontend/src/testing/__mocks__/implementations/matchMedia.mock.ts:4]: matchMedia was used by mantine

File: packages/frontend/src/testing/__mocks__/modules/ReactMarkdwnPreview.mock.tsx [hunk 15]
```diff
@@ -0,0 +1,11 @@
+import { FC } from 'react';
+
+type Props = {
+    source: string;
+};
+
+const ReactMarkdownPreview: FC<Props> = ({ source }) => {
+    return <>{source}</>;
+};
```
Comment (IrakliJani) [inline packages/frontend/src/testing/__mocks__/modules/ReactMarkdwnPreview.mock.tsx:9]: markdown preview mock is simple. just takes whatever and renders it without converting it to markdown. so if you search for an error message in a toaster - it should work!

File: packages/frontend/src/testing/__mocks__/providers/AppProvider.mock.tsx [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+            return {
+                ...userResponse,
+                ability: new Ability(userResponse.abilityRules),
+            };
+        },
+        enabled: !!health.data?.isAuthenticated,
+    });
+
+    return (
+        <AppProviederContext.Provider value={{ health, user }}>
```
Comment (IrakliJani) [inline packages/frontend/src/testing/__mocks__/providers/AppProvider.mock.tsx:44]: this file mocks real AppProvider by exposing mocked user and health data. it also accepts overrides as properties. that's how `renderWithProviders` and `renderHookWithProviders` produce mocked render.

File: packages/frontend/src/testing/vitest.setup.ts [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+
+    vi.mock('@uiw/react-markdown-preview', () => ({
+        default: ReactMarkdownPreview,
+    }));
+
+    mockMatchMedia();
+});
+
+// Disable all network requests by default
+nock.disableNetConnect();
```
Comment (IrakliJani) [inline packages/frontend/src/testing/vitest.setup.ts:19]: important.
disables all requests to the network!

File: packages/frontend/src/testing/vitest.setup.ts [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+
+afterEach(() => {
+    // Unmounts React trees that were mounted with render.
+    cleanup();
+
+    // Check if all nock interceptors were used
+    if (!nock.isDone()) {
+        console.table(
+            nock.pendingMocks().map((mock) => ({ 'Pending Mocks': mock })),
+            ['Pending Mocks'],
```
Comment (IrakliJani) [inline packages/frontend/src/testing/vitest.setup.ts:40]: print nicely what mocked (nocked?) requests were missed.

**Review by magnew — APPROVED** [review]
This looks good to me. I think maybe I'm missing some of the main motivation -- the goal is to make it possible to do more component testing **instead of using cypress**? Which seems fine to me. I do think we need better FE test coverage. This seems like one pretty good path to getting there. What do you think the plan should be for getting more coverage like this? Component tests for new complex components? What should still be tested with cypress?

I think my only real concern with this approach is producing and maintaining a lot of mocks can be very painful. But Im not sure it's less painful that fighting with cypress in ci + locally 😄. 

Nice work, I think this will be valuable.

File: packages/frontend/src/providers/AppProvider.tsx [hunk 19]
```diff
@@ -9,7 +9,9 @@ interface AppContext {
     user: UseQueryResult<UserWithAbility, ApiError>;
 }
 
-const Context = createContext<AppContext>(undefined as any);
+// used in test mocks
+// ts-unused-exports:disable-next-line
+export const AppProviederContext = createContext<AppContext>(undefined as any);
```
[the remark this answers is not in the document]
Comment (IrakliJani) [reply packages/frontend/src/providers/AppProvider.tsx:14]: good catch!

**Pull request merged**